packages/web-console/src/scenes/Editor/Monaco/index.tsx

```diff
@@ -309,8 +324,29 @@ const MonacoEditor = () => {
         },
       })
 
+      editor.onDidChangeModelContent((e) => {
+        // Remove errors when the invald query is updated
+        if (errorQueryHasFocusRef.current) {
```

Comment:
@argshook Thanks for the feedback. I agree that keeping the error markers until the error itself is touched is a better UX. 

One issue at the moment though is that the line decoration gets out of sync with error markers for multiline queries. In the screenshot below, the `hasError` class gets removed prematurely from the line decoration when the range check is implemented.

![multiline-error](https://user-images.githubusercontent.com/383846/235697633-7e4b0853-9564-40d7-9b7c-ef7be72a3e17.gif)

It looks like `renderLineMarkings` would need a few changes to keep the line decorations in sync for multiline queries. Could this be addressed in a follow-up? I'd be happy to open a new issue.

Reply:
> I agree that keeping the error markers until the error itself is touched is a better UX.

in the screencast seems like you already did the `if change within error`, right? If yes, then push it! :)

> It looks like `renderLineMarkings` would need a few changes to keep the line decorations in sync for multiline queries. Could this be addressed in a follow-up? I'd be happy to open a new issue.

seems like this issue is already apparent regardless whether it's a single or multiline query and despite the changes in this PR.

so yep, opening a dedicated issue would be appreciated. It can be
tackled in a separate PR.

thank you!

Reply:
After more testing, the "change within error" approach doesn't work as expected if the user first makes changes that affect the position of the error. In this case, the `errorRangeRef` no longer reflects the new position of the error text, and checking that edits are within the error range fail.

Working case when editing within the error range:
![error-range](https://user-images.githubusercontent.com/383846/235817257-287fbe50-107b-46cd-b6e2-0797b38a9dc9.gif)

Failing case when error position has drifted from error range ref:
![error-range-bug](https://user-images.githubusercontent.com/383846/235817277-d33e07f1-5f9d-446e-8f1c-3865763e3beb.gif)

My thought at the moment is to update the error range on content change, which looks possible but maybe not straightforward unless there is an existing method like `getErrorRange` that I'm overlooking. 

In hindsight, I think this case was part of my motivation for the "optimistic" approach, where the assumption is that the user intends to correct the error when editing any part of the invalid query. What do you think about going ahead with the optimistic approach and then revisit editing within the error as an improvement?

Reply:
> My thought at the moment is to update the error range on content change, which looks possible but maybe not straightforward unless there is an existing method like getErrorRange that I'm overlooking.

`getErrorRange` exists [here](https://github.com/conceptlogic/questdb-ui/blob/b373d7ca30219f07f39aa32c435ef394a48cc36d/packages/web-console/src/scenes/Editor/Monaco/utils.ts#L206) and it's used to set `errorRangeRef`, so this data is already available.

But i understand it's tricky, because to properly evaluate error range it must be done after every model change. However, not sure this can be done now, because real evaluation happens in the questdb server (after clicking `run`), so user must run the query for UI to know if it's erroneous or not.

Reply:
I'd love to see a few more test cases added (nothing grandiose, just a bit more coverage for more use cases). And then IMO it's good to merge.


wdyt @insmac ?